Java/core/src/main/java/com/amazon/randomcutforest/store/PointStore.java

```diff
@@ -290,135 +302,523 @@ public int getDimensions() {
+        return indexManager.capacity - indexManager.freeIndexPointer - 1;
+    }
+
     /**
      * a simple optimization that identifies the prefix of the arrays (refCount,
      * referenceList) that are being used
      * 
      * @return size of initial prefix in use
      */
     public int getValidPrefix() {
-        int prefix = capacity;
-        while (prefix > 0 && !occupied.get(prefix - 1)) {
+        int prefix = indexManager.capacity;
+        while (prefix > 0 && !indexManager.occupied.get(prefix - 1)) {
             prefix--;
         }
         return prefix;
     }
 
+    /**
+     * copy function for the store
+     * 
+     * @param dest   location to move to
+     * @param source moving from
+     * @param length number of values copied
+     */
     abstract void copyTo(int dest, int source, int length);
 
     /**
      * The following function eliminates redundant information that builds up in the
-     * pointstore and shrinks the pointstore
+     * point store and shrinks the point store
      */
 
     public void compact() {
-        checkArgument(!directLocationMap, "incorrect call; should not be used for direct location maps");
-        int DEFAULT_EMPTY = 0; // indicates location not in use
 
         int runningLocation = 0;
         startOfFreeSegment = 0;
-        int stepDimension = (shingleAwareOverlapping) ? baseDimension : dimensions;
-
-        // we first determine which locations are the startpoints of the shingles
-        int[] reverseReferenceList = new int[capacity * dimensions / stepDimension];
-        for (int i = 0; i < capacity; i++) {
-            if (occupied.get(i)) {
-                reverseReferenceList[locationList[i] / stepDimension] = i + 1;
-                // offset by 1, to distinguish DEFAULT_EMPTY
+        int stepDimension = baseDimension;
+
+        // we first determine which locations are the start points of the shingles
+        // since the shingles extend for a length and can overlap this help define the
+        // region that should be copied
+
+        HashMap<Integer, Integer> movedTo = new HashMap<>();
+
+        // the bit set corresponds to the locations that can be in use over the actual
+        // store array
+        // this is not the same as the number of points that can be stored
+        BitSet inUse = new BitSet(indexManager.capacity);
+
+        for (int i = 0; i < indexManager.capacity; i++) {
+            if (indexManager.occupied.get(i)) {
+                validateInternalState(refCount[i] > 0, " bits out of sync when count = 0 ");
+                validateInternalState(locationList[i] != PointStore.INFEASIBLE_POINTSTORE_INDEX,
+                        " incorrect location info");
+                inUse.set(locationList[i] / stepDimension);
+            } else {
+                validateInternalState(refCount[i] == 0, "bits out of sync with count > 0 ");
+                locationList[i] = INFEASIBLE_LOCATION;
             }
+
         }
```

Comment:
Looks like we could separate this out into a protected method and then write tests for it directly.

Reply:
Done